sdk/core/allocator/revoker.h

```diff
@@ -87,41 +99,58 @@ namespace Revocation
 		 */
 		void shadow_paint_range(ptraddr_t base, ptraddr_t top, bool fill)
 		{
-			constexpr size_t ShadowWordAddrMask =
-			  (1U << (ShadowWordShift + MallocAlignShift)) - 1;
-			ptraddr_t baseUp =
-			  (base + ShadowWordAddrMask) & ~ShadowWordAddrMask;
-			ptraddr_t topDown = top & ~ShadowWordAddrMask;
-
-			// There isn't a single aligned shadow word for this range, so paint
-			// one bit at a time.
-			if (baseUp >= topDown)
+			size_t baseCapOffset = (base - TCMBaseAddr) >> MallocAlignShift;
+			size_t topCapOffset  = (top - TCMBaseAddr) >> MallocAlignShift;
+			size_t baseWordIx    = baseCapOffset >> ShadowWordShift;
+			size_t topWordIx     = topCapOffset >> ShadowWordShift;
+
+			WordT maskLo = shadow_paint_bits_above(base);
+			WordT maskHi = shadow_paint_bits_below(top);
+
+			if (baseWordIx == topWordIx)
 			{
-				for (ptraddr_t ptr = base; ptr < top; ptr += MallocAlignment)
+				/*
+				 * This object is entirely contained within one word of the
+				 * bitmap.  We must AND the mask{Hi,Lo} together and use that
+				 * to update the right bits.
+				 */
+				WordT mask = maskHi & maskLo;
+				if (fill)
+				{
+					shadowCap[baseWordIx] |= mask;
+				}
+				else
 				{
-					shadow_paint_single(ptr, fill);
+					shadowCap[baseWordIx] &= ~mask;
 				}
+
 				return;
 			}
 
-			// First, paint the individual bits at the beginning.
-			for (ptraddr_t ptr = base; ptr < baseUp; ptr += MallocAlignment)
+			/*
+			 * Otherwise, there are at least two words of the bitmap that need
+			 * to be updated with the masks, and possibly some in between that
+			 * just need to be set wholesale.
+			 */
+
+			WordT midWord;
+			if (fill)
 			{
-				shadow_paint_single(ptr, fill);
+				shadowCap[baseWordIx] |= maskLo;
+				shadowCap[topWordIx] |= maskHi;
+				midWord = ~0;
```

Comment:
```suggestion
			midWord = ~WordT(0);
```

You are currently relying on `WordT` being smaller than `int` or signed, implicitly.